x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/datafeed/DatafeedConfig.java

```diff
@@ -799,9 +801,10 @@ public Builder(StreamInput in) throws IOException {
                 this.indices = null;
             }
             // each of these writables are version aware
-            this.queryProvider = QueryProvider.fromStream(in);
+            this.queryProvider = in.readOptionalWriteable(QueryProvider::fromStream);
             // This reads a boolean from the stream, if true, it sends the stream to the `fromStream` method
             this.aggProvider = in.readOptionalWriteable(AggProvider::fromStream);
+            this.esqlQuery = in.readOptionalString();
```

Comment:
As above, use checks on `TransportVersion` to make the serialization compatible with old nodes.